client/tracing/src/block/mod.rs

```diff
@@ -0,0 +1,314 @@
+// Copyright 2021 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Utilities for tracing block execution
+
+use std::{collections::HashMap, sync::{Arc, atomic::{AtomicU64, Ordering}}};
+use std::time::Instant;
+
+use parking_lot::Mutex;
+use tracing::{Dispatch, dispatcher, Subscriber, Level, span::{Attributes, Record, Id}};
+use tracing_subscriber::CurrentSpan;
+
+use sc_client_api::BlockBackend;
+use sc_rpc_server::MAX_PAYLOAD;
+use sp_api::{Core, Metadata, ProvideRuntimeApi, Encode};
+use sp_blockchain::HeaderBackend;
+use sp_runtime::{
+	generic::BlockId,
+	traits::{Block as BlockT, Header},
+};
+use sp_rpc::tracing::{BlockTrace, Span, TraceError, TraceBlockResponse};
+use sp_tracing::{WASM_NAME_KEY, WASM_TARGET_KEY, WASM_TRACE_IDENTIFIER};
+use sp_core::hexdisplay::HexDisplay;
+use crate::{SpanDatum, TraceEvent, Values};
+
+// Estimate of the max size base RPC payload when the Id is bound as a u64. If strings
+// are used for the RPC Id this may need to be adjusted. Note: The base payload does not
+// include the RPC result.
+const BASE_RPC_PAYLOAD: usize = 500;
```

Comment:
You may want the clarify that we estimate the max `base size` of the JSONRPC response message and it has the following format:

`{"jsonrpc":"2.0","result":[],"id":18446744073709551615}`

If we ignore huge strings as IDs it can only be 55 bytes, https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=85558e03fb6f28dcf36484d7f9b6714c but 500 is probably fair ^^

Reply:
>  https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=85558e03fb6f28dcf36484d7f9b6714c
This is a cool example!